add M220 speed factor and M221 flow factor

hg42 opened the pull request:
implements #50 

* for consistency both factors are handled similar in the gcode module.
* they are stored and used in each G0/G1 as parameter for toolhead.move() which creates a Move object with it
* Move.__init__ handles both factors.
* speed_factor is trivial
* flow_factor is the reason why the handling is done in Move, because here the end position of a move is converted into a distance. Handling the end position and current position with a scaling factor in upper levels, is way too complex and error prone.

hg42 commented:
note, I cannot test this with the master branch, because my machine can currently only run on my own branch (XY on separate MCUs and endstops behavior).

So, while tested on my machine, the extracted changes for this PR are untested in this branch.
Please test yourself.
This PR could also be delayed until other patches are incorporated.

KevinOConnor commented:
This is fine for testing, but I don't think it is suitable for the mainline code.  If I'm understanding the functionality correctly, a change in flowrate or speed would be used during calibration, and it is likely to be used a handful of times a year.  The toolhead.move() command is often called on an order of 100 times a second.  I'd prefer to keep the toolhead.move() code path optimized to the common case (for both cpu efficiency and code maintenance).

Some notes:
- the mcu is indirectly programmed with the max velocity (see mcu.py:setup_min_stop_interval() ) - if the max velocity is exceeded via a speed change then it could lead to MCU shutdowns.
- the current extruder position is cached within the gcode, toolhead, extruder, mcu_stepper, and stepcompression code.  If the code changes the position (via a flowrate change) then it must resynchronize all of these classes to the new position.  Otherwise, a step compression error is likely to be raised.

Cheers,
-Kevin

KevinOConnor commented:
Oh - as an additional note - my preference would be to use "extended g-code" commands for this type of functionality.  As in "SET_FLOWRATE" and "SET_SPEED_FACTOR" instead of M220/M221.  I find g-code cryptic and hard to use - my goal is to only support g-code commands that are in use by other widely available software tools.

hg42 commented:
M220/M221 are pretty standard

http://reprap.org/wiki/G-code#M220:_Set_speed_factor_override_percentage

They are at least available in the user interface of Repetier Host and Printrun.

I (and everyone I know) use both on most prints. It's part of fine tuning, adjusting the flow when the skirt/brim and the first layers are printed. You need speed controlwhen printing very fast.
You can hear, if you are near the limits and slow down a bit. Also some prints trigger resonances, which you can easily reduce by only a slight change in speed.
Actually, I would want to have different speeds and flows for each of infill, inner shells, surfaces, etc. Unfortunately this is not possible without being marked by the slicer (but many do).

hg42 commented:
I can understand your preference for clear text commands, actually I would like that, too.

Unfortunately in reality gcode is a standard and users use tools to process gcode files after slicing. If these textual commands are included in start, end, tool-change gcode etc., many tools will choke on those lines.
Thus, having such commands for interactive use is fine, but it's better to have alternative forms that obey gcode standards.

Tools like octoprint try stay gcode compliant, e.g. they use a syntax like `//action=foobar` for extensions (which I think is still not compliant, but seems to work with usual firmware).

According to http://reprap.org/wiki/G-code RepRap follows this standard: http://ws680.nist.gov/publication/get_pdf.cfm?pub_id=823374

There are many things that only a few people know:
* "an optional block delete character, which is a slash “/” "
* "Any input not explicitly allowed is illegal and will cause the Interpreter to signal an error"
* "Spaces and tabs are allowed anywhere on a line of code and do not change the meaning of the
line, except inside comments. This makes some strange-looking input legal. The line “g0x +0. 12
34y 7” is equivalent to “g0 x+0.1234 y7”, for example"
* "A word is a letter other than N followed by a real value"
* "Words may begin with any of the letters shown in Table 3"
* "The first group (the words) may be reordered in any way without changing the meaning of the
line" which means you can write X0 G1 Y0
* there are comments like "G1 X0 (this is X) Y0 (this is Y)"

There are also interesting features like expressions and a kind of variables.
The document defines only a few M-codes.

As a result usual gcode processors will only look at the first char and read a number after that, then read the next char, then a number etc. Even in reprap ecosystems.

I bet very few of the rules from that document are obeyed by usual firmwares (e.g. I guess they will not interpret something like "X0 Y0 G1"), but because it's kind of a subset of the syntax, gcode given to them is usually valid.

A usual solution is to format commands as comments with a special start sequence, but I doubt that usual reprap firmwares know of "(...)" notation (but I never tried).

hg42 commented:
> I'd prefer to keep the toolhead.move() code path optimized to the common case (for both cpu efficiency and code maintenance).

move is nothing else like G0 or G1. I don't see another possibility but one simple multiplication.

> the mcu is indirectly programmed with the max velocity (see mcu.py:setup_min_stop_interval() ) - if the max velocity is exceeded via a speed change then it could lead to MCU shutdowns

max_velocity could be checked, but M220 does not know which moves will come in. So this check would be necessary for each move command.
Or do you mean to also increase max_velocity.
But it is in no way different from incoming speeds that are too high. Klipper should limit the speeds and not stop printing.

> the current extruder position is cached within the gcode, toolhead, extruder, mcu_stepper, and stepcompression code

this is exactly why I implemented it at this place and not at the gcode level (which would be possible otherwise but also much more complicated).
I chose the point after the current and target positions are converted into a distance. If a lower level would cache the position it would be different from that of a higher level, but correct as the real physical position, while the upper levels use the virtual position of the gcode file. I assume, the code does not read back this physical position to the higher level at any place, which would be plain wrong in my eyes.

laudiblellc commented:
I would comment also that the ability to modify speed and feed on the fly is a basic feature and should be incorporated. I've wanted it many times already! Octoprint already has the interface built in, so that saves some work.

KevinOConnor commented:
On Thu, Dec 21, 2017 at 03:28:15AM -0800, Harald wrote:
> > the mcu is indirectly programmed with the max velocity (see mcu.py:setup_min_stop_interval() ) - if the max velocity is exceeded via a speed change then it could lead to MCU shutdowns
> 
> max_velocity could be checked, but M220 does not know which moves will come in. So this check would be necessary for each move command.

Right.  Unless I missed something, the code pull request moved the
multiplication to after Klipper's existing speed check (in
toolhead.move() ) - thus defeating the existing check.

> > the current extruder position is cached within the gcode, toolhead, extruder, mcu_stepper, and stepcompression code
> 
> this is exactly why I implemented it at this place and not at the gcode level (which would be possible otherwise but also much more complicated).

Unless I missed something, the pull request did not implement this
correctly.  The move class stores start and end position in absolute
coordinates with axes_d being the difference between the two
positions.  If these values don't line up correctly then the
subclasses will get out of sync on their perceived positions and it
will result in step compression errors.

-Kevin

KevinOConnor commented:
On Thu, Dec 21, 2017 at 01:52:53PM +0000, laudiblellc wrote:
> I would comment also that the ability to modify speed and feed on the fly is a basic feature and should be incorporated. I've wanted it many times already! Octoprint already has the interface built in, so that saves some work.

Can you point me to where I can find this in Octoprint?

Thanks,
-Kevin

hg42 commented:
![image](https://user-images.githubusercontent.com/4707597/34269059-40bb9c96-e683-11e7-82b5-fdf2ae0617d1.png)

KevinOConnor commented:
Hah - I've never actually noticed that before.  So, yes, I agree this
will need to be supported in Klipper.

I would prefer to see this implemented entirely in the gcode layer
however.  If I get time I'll check if that is feasible.

-Kevin

hg42 commented:
> Unless I missed something, the code pull request moved the multiplication to after Klipper's existing speed check (in toolhead.move() ) - thus defeating the existing check.

yes, you are right, I missed that check.

Initially I implemented speed_factor in gcode.py.
But because flow_factor has to be applied to the difference and cannot change the positions related to the gcode, it had to be moved downward at least to toolhead.py.
Then I thought it would be better to move speed_factor to the same place, to have it more consistent.

I also didn't expect the speed check on an upper level, because I would check it just before generating steps and there is limit_speed in Move and check_move in extruder etc.
You never know which algorithm changes speeds in between.

I see the positions are used in extruder and kinematics.
But end_pos is not used in extruder and that's why it works. But I agree this was lucky, I should have checked which values are stored in Move and also adjust end_pos.
(end_pos is used in DummyExtruder for an exception value, but it is not checked there, this doesn't seem to make sense...)

I moved speed_factor to toolhead.move() and added adjustment of end_pos[3] even if it is not used. I will start testing, when I designed some objects to be printed.

Some more thoughts:
* kinematic classes may need a base class and duplicated code could go there
* step ramp generation is also copied to extruder.py, this could probably be moved to class Move or a helper function.

hg42 commented:
> I would prefer to see this implemented entirely in the gcode layer however. If I get time I'll check if that is feasible

I would generally prefer this too, but it is very difficult to do this right for flow control.
The flow_factor theoretically belongs to steps_per_mm.

As an indicator, several move commands like arcs etc. would have to be handled in the gcode layer. But they all go through toolhead.move() or even better a lower layer.
Actually, algorithms like pressure_advance should be applied before flow correction. It's only lucky that it works both ways, before or after (I hope).

hg42 commented:
if you calibrate extruder steps/mm with a kind of standard filament (you have to choose one), you can apply a different flow factor for each filament spool. When changing the spool, you do not need to configure and restart the system, but simply change the flow factor. The slicer settings can also be kept at least for the same kind of material, if other parameters like retraction etc. do not need a change.

I usually adjust the flow visually by looking at the first layers after the bottom layer. I even adjust the bed leveling on the bottom layer if necessary. And if you see the adhesion doesn't work well, you can reduce the speed factor, which helps a lot. That is why I tend to use several rounds of skirt.

KevinOConnor commented:
On Thu, Dec 21, 2017 at 08:02:46PM +0000, Harald wrote:
> I moved speed_factor to toolhead.move() and added adjustment of end_pos[3] even if it is not used.

FYI, move1.end_pos must be equal to move2.start_pos.

>I will start testing, when I designed some objects to be printed.

Take a look at the "Translating gcode files" section of
docs/Debugging.md - it's a good way to unit test changes without
requiring hardware.

-Kevin

KevinOConnor commented:
On Thu, Dec 21, 2017 at 08:22:03PM +0000, Harald wrote:
> > I would prefer to see this implemented entirely in the gcode layer however. If I get time I'll check if that is feasible
> 
> I would generally prefer this too, but it is very difficult to do this right for flow control.

I took a look and I don't think it is that bad - see the
work-m221-20121222 branch I just pushed (unit tested only).

-Kevin

hg42 commented:
ok, this looks correct...
I didn't think of the possibility to scale the coordinate system for E while anchoring it at last_position.
Otherwise the stored current positions in lower levels would make this fail. That's why I immediately dropped that.
(btw. I don't like the name speed_factor, when it's two things in one: speed_factor and scaling min->sec)

hg42 commented:
I "dry" test worked.
I'll do a real test tomorrow.

hg42 commented:
the wet test works, too.

It would be better if E speed (retraction!) wouldn't be affected my M220.
The nozzle stays too long at the same spot, because retraction is so slow, which burns the layer and blobs.
I generally think speed could be separated, something like M220 X200 Y150 Z100 E150 would be useful. I would apply Sxxx only to XY.

KevinOConnor commented:
On Sat, Dec 23, 2017 at 08:15:06PM +0000, Harald wrote:
> the wet test works, too.

Thanks!

> It would be better if E speed (retraction!) wouldn't be affected my M220.
> The nozzle stays too long at the same spot, because retraction is so slow, which burns the layer and blobs.
> I generally think speed could be separated, something like M220 X200 Y150 Z100 E150 would be useful. I would apply Sxxx only to XY.

I don't understand - the retraction speed should be dominated by the
configured max_extrude_only_velocity.

FYI, you've raised a number of items across a number of different
topics.  I'm struggling to keep track of them.  It will likely be
several days before I'll be able to respond.

-Kevin

hg42 commented:
> the retraction speed should be dominated by the configured max_extrude_only_velocity

this limits the maximum.
But I am talking about the slow down if you do something like `M220 S25`.

You usually want to slow down XY and keep retraction speed, because retraction has nothing to do with motion speeds.

But being able to give a factor for each axis separately would be even better. There may be cases where you only want to speed up the extruder. Or because X and Y often have different masses, you may want to only slow down Y.

Sorry Kevin, I do not want to push you. Just take the time you need.
